lib/unison-util-recursion/src/Unison/Util/Recursion.hs

```diff
@@ -0,0 +1,57 @@
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Unison.Util.Recursion
+  ( Algebra,
+    Recursive (..),
+    cataM,
+    para,
+    Fix (..),
+    Cofree' (..),
+  )
+where
+
+import Control.Arrow ((&&&))
+import Control.Comonad.Cofree (Cofree ((:<)))
+import Control.Monad ((<=<))
+
+type Algebra f a = f a -> a
+
+class Recursive t f | t -> f where
```

Comment:
Just curious why not use `recursion-schemes` directly?

I personally find your encoding here less confusing than all the `Base` type-family nonsense, but it's also nice to maintain compatibility with the ecosystem where possible 😄

Reply:
Yeah, my plan was to replace this with a recursion scheme library eventually. At this point, I was just trying to centralize the (mostly) existing copies of this pattern.

I’m happy to put in recursion-schemes, but I also have my own recursion scheme library, [Yaya](https://hackage.haskell.org/package/yaya), which makes some different decisions than recursion-schemes[^1]. One of them is the use of `FunctionalDependencies`, as you pointed out. Another is that Yaya tries to provide totality (especially when paired with [`NoRecursion`](https://hackage.haskell.org/package/no-recursion)).

So I just punted on that discussion for now.

[^1]: There’s a more in-depth comparison with recursion-schemes in [Yaya’s README](https://github.com/sellout/yaya/blob/main/core/README.md#recursion-schemes).